### DISCUSSION TITLE
strange behaviour of nlp.tokenizer.prefix_search with digits

### DISCUSSION THREAD

**Discussion by DSLituiev** [explosion/spaCy#8218] [Help: Coding & Implementations] [answered]
I'm running into a mysterious treatment of digit "8" by the `nlp.tokenizer.prefix_search`:

```python
import spacy
spacy.__version__
# '3.0.5'

nlp = spacy.load("en_core_web_sm")
prefixes = nlp.Defaults.prefixes + [
                                    r"/", # '\.',
                                    r'''^[±\-\+., ]*[0-9]+(\.[0-9]+)?''' 
                                   ]
prefix_regex = spacy.util.compile_prefix_regex(prefixes)
nlp.tokenizer.prefix_search = prefix_regex.search

nlp.tokenizer.url_match = None
```

test:
```
list(nlp(" (0/8). (0/9). (1/5)"))
# [ , (, 0, /, 8), ., (, 0, /, 9, ), ., (, 1, /, 5, )]
```

for some reason digit "8" keeps grouped the ")", while other digits aren't. What can be happening here?

**Comment by polm** [comment, marked as answer]
Looks like `8)` is getting picked up as a smiley. I'll see if I can dig it up in the code.

**Reply by polm** [reply]
There might be a better way to disable emoticons in general if you don't need them, but this seems to work as a quick fix:

```
import spacy
del spacy.lang.tokenizer_exceptions.BASE_EXCEPTIONS["8)"]

nlp = spacy.blank("en")
# now "8)" is the same as "9)"
```

**Reply by DSLituiev** [reply]
This doesn't seem to work:

```python
spacy.__version__
# '3.0.6'

try:
    del spacy.lang.tokenizer_exceptions.BASE_EXCEPTIONS["8)"]
except:
    print("no 8)")
# nlp.tokenizer._flush_specials()

nlp = spacy.load("en_core_web_sm")
list(nlp(" (0 / 8). (0 / 9). (1 / 5)"))
# [ , (, 0, /, 8), ., (, 0, /, 9, ), ., (, 1, /, 5, )]

**Reply by DSLituiev** [reply]
Note: I'm using `"en_core_web_sm"`, not `"en"` as the base model

**Reply by polm** [reply]
This can be fixed with a tokenizer exception (which is probably what I should have done to begin with), like so:

```
import spacy
nlp = spacy.load("en_core_web_sm")

nlp.tokenizer.add_special_case("8)", [{"ORTH": "8"}, {"ORTH": ")"}])
print(list(nlp("( 0 / 8).")))
# => [(, 0, /, 8, ), .]
```

See [the tokenizer exceptions docs](https://spacy.io/usage/linguistic-features#special-cases) for more info on that.

> Note: I'm using "en_core_web_sm", not "en" as the base model

`spacy.blank("en")` loads a tokenizer with no model (note it's `blank`, not `load`), so `en` is just the language name. 

I believe my original fix didn't work with the model because the model uses a serialized tokenizer.

**Reply by adrianeboyd** [reply]
As a note, it would be bit cleaner to just delete the exception:

```python
nlp.tokenizer.rules = {key: value for key, value in nlp.tokenizer.rules.items() if key != "8)"}
```

(Reassigning the complete exceptions dict to the `nlp.tokenizer.rules` property makes sure that all the internal tokenizer caches are cleared.)
